### ISSUE TITLE
Comparison of bigsnpr versions 1.11.4  and 1.10.8

### ISSUE THREAD

**Issue by nathangillespie** [privefl/bigsnpr#390]
Dear Authors,
We are using LDpred2 in bigsnpr to estimate weights to generate polygenic risk scores. Here, we note the much lower than expected correlations between weights generated using different versions of bigsnpr across three servers (see below). The correlation between beta weights using bigsnpr version 1.11.4  and 1.10.8 is 0.55. What’s more, a PRS for education attainment using the EA4 sumstats and bigsnpr 1.11.4 is no longer predictive of EA in an independent sample, whereas the EA4 PRS when based on the older bigsnpr 1.10.8 weights remains significant. 
Is a correlation of 0.55 between output from adjacent versions reasonable?
Regards,
Nathan 

```
package:       VCU       Dutch       TW
R 4.1.1       4.0.3       3.6.0
bigsnpr        1.11.4     1.11.4     1.10.8
bigreadr        0.2.4       0.2.4       0.2.4
```

VCU 1.11.4  & VCU 1.10.8       
bigsnpr package r = 0.55

VCU 1.10.8 and TW 1.10.8
bigsnpr package r = 1.00

Dutch 1.10 and Dutch  v1.11
bigsnpr package r= 0.55

**Comment by privefl** [comment]
I am a bit surprised that you get a 100% correlation by running LDpred2 twice.
There is some random sampling, so unless you use seeds, two runs of LDpred2 should give slightly different results.

But indeed r=0.55 is quite small, and that might reveal some issue.
I will have time to investigate this further next week.
In the meantime, could you please report the `range` that you get with the two versions?

**Comment by privefl** [comment]
I've tried the EA4 sumstats (N=765K), and it seems the new model is better for it.
At least `range` and predicted r2 from the Gibbs sampler are larger.

Maybe you could send me your code and I could try it.

**Comment by nathangillespie** [comment]
Hi Mei-Hsin, Can you please send the developers the code you have been using for running the comparisons? Cheers, Nathan

On Nov 29, 2022, at 11:22 AM, Florian Privé ***@***.***> wrote:



I've tried the EA4 sumstats (N=765K), and it seems the new model is better for it.
At least range and predicted r2 from the Gibbs sampler are larger.

Maybe you could send me your code and I could try it.

**Comment by nathangillespie** [comment]
Below is the code we have used cross the 3 versions and the ranges:

![range](https://user-images.githubusercontent.com/67481270/204635354-2d660f6e-0f06-4ee3-8062-31bf32f2079a.png)



# prepare workspace and load bigsnpr
	# remove.packages("bigsnpr")
	# library(remotes)
	# install_version("bigsnpr", "1.10.8")
	# install_version("bigsnpr", "1.11.6")
library(bigsnpr)
options(bigstatsr.check.parallel.blas = FALSE)




#==================== 1. Import LD reference ====================
# position is in build GRCh37 / hg19, but we provide positions in 3 other builds 
info <- readRDS("map.rds")
write.table(info, file = "map.txt", quote= FALSE,row.names=FALSE,col.names=FALSE)


str(info)







#==================== 3. Import summary statistics ====================
sumstats <- bigreadr::fread2("EA4_additive_excl_23andMe_serverVtest.txt")
str(sumstats)

# Adjust header names
# Need to be as specified in https://urldefense.com/v3/__https://privefl.github.io/bigsnpr/articles/LDpred2.html__;!!JqxBPMk!mM86lOe54fBj1RFIl1E3Hg18VZcEMgxaIBSv10Pm5RQEzhVwt5ySkQ8Afc8lhfx7nCOluhmuCL32e5LtKY67Tfzm$ 
# a0 is the effect allele
names(sumstats) <-
    c("rsid",
    "chr",
    "pos",
    "a0",
    "a1",
    "EAF_HRC",
    "beta", 
    "beta_se_adj",
    "beta_se",
    "p_adj",
    "p")

sumstats$n_eff <- 765283 

# Get overlap of SNPs in LDref and in sumstats
df_beta <- snp_match(sumstats, info)
str(df_beta)

	


#==================== 4. Compute LD matrix ====================

# (1) compute correlations between variants
corr <- NULL

# (2) Set amount of cores to be used and disable parallel check
NCORES <- 1

# (3) Temporary directory for storage of matrix
tmp <- tempfile(tmpdir = "/.../v6_bigsnpr1.11.6")

# (4) Correlation matrix 
for (chr in 1:22) {

  cat(chr, ".. ", sep = "")
  ind.chr <- which(df_beta$chr == chr)
  ind.chr2 <- df_beta$`_NUM_ID_`[ind.chr]
  ind.chr3 <- match(ind.chr2, which(info$chr == chr))
  corr_chr <- readRDS(paste0("/.../LD_with_blocks_chr", chr, ".rds"))[ind.chr3, ind.chr3] 
 
  if (chr == 1) {
    corr <- as_SFBM(corr_chr, tmp)
  } else {
    corr$add_columns(corr_chr, nrow(corr))
  }
}



#==================== 5. Calculate weight ====================
# (1) Perform LD score regression
ld <- NULL
NCORES <- 1

(ldsc <- with(df_beta, snp_ldsc(ld, length(ld), chi2 = (beta / beta_se)^2,
                                sample_size = n_eff, blocks = NULL)))

h2_est <- ldsc[["h2"]] 

# (2) Run model for LDpred2-auto 
multi_auto <- snp_ldpred2_auto(corr, df_beta, h2_init = h2_est,
                               vec_p_init = seq_log(1e-4, 0.2, length.out = 30),
                               allow_jump_sign = FALSE, shrink_corr = 0.95,
                               ncores = NCORES)

# (3) filtering bad chains (by comparing the scale of the resulting predictions)
range <- sapply(multi_auto, function(auto) diff(range(auto$corr_est)))
keep <- (range > (0.9 * quantile(range, 0.9))) 

# (4) Apply filter to get final adjusted weights
beta_auto <- rowMeans(sapply(multi_auto[keep], function(auto) auto$beta_est))

# (5) Save SNP weights ###

weights <- data.frame(df_beta$rsid,df_beta$a0,beta_auto)
write.table(weights, file = "/.../EA4_weights_bigsnpr1.11.6.txt", quote= FALSE,row.names=FALSE,col.names=FALSE)
# save.image(file = "VCU.RData") # It will save every object that you create

**Comment by nathangillespie** [comment]
Here are the ranges:
![range](https://user-images.githubusercontent.com/67481270/204635600-dd97887b-47ca-48f4-a345-eb93d6d16084.png)

**Comment by nathangillespie** [comment]
Here is a link to the R script we have been using:

[https://drive.google.com/file/d/1a1xglvaKH7d_x3A2LZft4RDL4odivvjo/view?usp=share_link](url)

**Comment by nathangillespie** [comment]
You're correct. The correlations between VCU 1.10.8 and TW 1.10.8 bigsnpr package were not perfect but actually 0.99.

**Comment by privefl** [comment]
The ranges are what I get too.
The estimated r2 is actually very large with the new version (19%) for me, so I'm quite surprised when you say that it is not predictive at all..

**Comment by privefl** [comment]
You're also missing the QC step on the sumstats.

**Comment by privefl** [comment]
If you install the newest GitHub version, you can now run LDpred2-auto as before when using `use_MLE = FALSE`.

**Comment by privefl** [comment]
Any update on this?
